File: frontend/src/services/certificate.service.ts
import api from "./api";

export interface CertificateResponse {
  id: number;
  certificateNumber: string;
  verificationCode: string;

  userId: number;
  contestId: number;

  participantName?: string;
  contestName?: string;

  type:
    | "FIRST_PLACE"
    | "SECOND_PLACE"
    | "THIRD_PLACE"
    | "PARTICIPATION";

  rank: number | null;

  issuedAt: string;

  certificateUrl: string | null;
}


/* =========================================================
   GET MY CERTIFICATE FOR A CONTEST
========================================================= */

export const getMyContestCertificate = async (
  contestId: number
): Promise<CertificateResponse> => {

  const response =
    await api.get(
      `/certificates/contest/${contestId}/my`
    );

  return response.data;
};


/* =========================================================
   DOWNLOAD CERTIFICATE PDF
========================================================= */

export const downloadCertificate = async (
  certificateId: number
): Promise<Blob> => {

  const response =
    await api.get(
      `/certificates/${certificateId}/download`,
      {
        responseType: "blob",
      }
    );

  return response.data;
};


/* =========================================================
   GET ALL MY CERTIFICATES
========================================================= */

export const getMyCertificates = async (): Promise<
  CertificateResponse[]
> => {

  const response =
    await api.get(
      "/certificates/my"
    );

  return response.data;
};

File: backend/src/main/java/com/corvuscode/certificate/service/CertificateService.java
package com.corvuscode.certificate.service;

import com.corvuscode.certificate.dto.response.CertificateResponse;
import com.corvuscode.certificate.entity.Certificate;

import java.util.List;

public interface CertificateService {

    Certificate generateCertificate(
            Long contestId,
            Long userId,
            Integer rank
    );

    void generateContestCertificates(
            Long contestId
    );

    List<CertificateResponse>
    getUserCertificates(
            Long userId
    );

    List<CertificateResponse>
    getContestCertificates(
            Long contestId
    );

    CertificateResponse
    getMyContestCertificate(
            Long contestId,
            String email
    );

    CertificateResponse
    verifyCertificate(
            String verificationCode
    );

    byte[] downloadCertificate(
            Long certificateId
    );

    List<CertificateResponse>
    getMyCertificates(
            String email
    );

    boolean isCertificateOwnedByUser(
            Long certificateId,
            String email
    );
}